
<font size="8"><center>**Heart Failure Analysis**</center></font>

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Group-information" data-toc-modified-id="Group-information-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Group information</a></span></li><li><span><a href="#Dataset-information" data-toc-modified-id="Dataset-information-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Dataset information</a></span><ul class="toc-item"><li><span><a href="#Subject-of-the-data" data-toc-modified-id="Subject-of-the-data-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Subject of the data</a></span></li><li><span><a href="#Source-of-the-data-and-how-this-dataset-collected?" data-toc-modified-id="Source-of-the-data-and-how-this-dataset-collected?-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Source of the data and how this dataset collected?</a></span></li><li><span><a href="#Copyright" data-toc-modified-id="Copyright-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Copyright</a></span></li></ul></li><li><span><a href="#Data-exploration" data-toc-modified-id="Data-exploration-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data exploration</a></span><ul class="toc-item"><li><span><a href="#Load-the-dataset" data-toc-modified-id="Load-the-dataset-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Load the dataset</a></span></li><li><span><a href="#Get-size-of-the-dataset" data-toc-modified-id="Get-size-of-the-dataset-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Get size of the dataset</a></span></li><li><span><a href="#Exploring-the-rows" data-toc-modified-id="Exploring-the-rows-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Exploring the rows</a></span><ul class="toc-item"><li><span><a href="#Meaning-of-each-rows" data-toc-modified-id="Meaning-of-each-rows-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Meaning of each rows</a></span></li><li><span><a href="#Check-the-duplicated-row" data-toc-modified-id="Check-the-duplicated-row-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Check the duplicated row</a></span></li></ul></li><li><span><a href="#Exploring-the-columns" data-toc-modified-id="Exploring-the-columns-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Exploring the columns</a></span><ul class="toc-item"><li><span><a href="#Meaning-of-the-columns" data-toc-modified-id="Meaning-of-the-columns-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Meaning of the columns</a></span></li><li><span><a href="#Datatype-of-each-columns" data-toc-modified-id="Datatype-of-each-columns-3.4.2"><span class="toc-item-num">3.4.2&nbsp;&nbsp;</span>Datatype of each columns</a></span></li><li><span><a href="#Explore-numeric-column(s)" data-toc-modified-id="Explore-numeric-column(s)-3.4.3"><span class="toc-item-num">3.4.3&nbsp;&nbsp;</span>Explore numeric column(s)</a></span></li><li><span><a href="#Explore-categorical-column(s)" data-toc-modified-id="Explore-categorical-column(s)-3.4.4"><span class="toc-item-num">3.4.4&nbsp;&nbsp;</span>Explore categorical column(s)</a></span><ul class="toc-item"><li><span><a href="#Explore-the-binary-attributes" data-toc-modified-id="Explore-the-binary-attributes-3.4.4.1"><span class="toc-item-num">3.4.4.1&nbsp;&nbsp;</span>Explore the binary attributes</a></span></li><li><span><a href="#Explore-object-attribute" data-toc-modified-id="Explore-object-attribute-3.4.4.2"><span class="toc-item-num">3.4.4.2&nbsp;&nbsp;</span>Explore object attribute</a></span></li></ul></li><li><span><a href="#Further-explore" data-toc-modified-id="Further-explore-3.4.5"><span class="toc-item-num">3.4.5&nbsp;&nbsp;</span>Further explore</a></span></li></ul></li></ul></li><li><span><a href="#Asking-meaningful-questions-and-answering-them" data-toc-modified-id="Asking-meaningful-questions-and-answering-them-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Asking meaningful questions and answering them</a></span><ul class="toc-item"><li><span><a href="#How-to-detect-people-who-have-heart-disease-by-those-symptoms?" data-toc-modified-id="How-to-detect-people-who-have-heart-disease-by-those-symptoms?-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>How to detect people who have heart disease by those symptoms?</a></span><ul class="toc-item"><li><span><a href="#Purposes" data-toc-modified-id="Purposes-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Purposes</a></span></li><li><span><a href="#Pre-processing" data-toc-modified-id="Pre-processing-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>Pre-processing</a></span></li><li><span><a href="#Analysis" data-toc-modified-id="Analysis-4.1.3"><span class="toc-item-num">4.1.3&nbsp;&nbsp;</span>Analysis</a></span></li></ul></li><li><span><a href="#Which-factors-leading-to-the-heart-disease?" data-toc-modified-id="Which-factors-leading-to-the-heart-disease?-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Which factors leading to the heart disease?</a></span><ul class="toc-item"><li><span><a href="#Purposes" data-toc-modified-id="Purposes-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Purposes</a></span></li><li><span><a href="#Pre-processing" data-toc-modified-id="Pre-processing-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>Pre-processing</a></span></li><li><span><a href="#Analysis" data-toc-modified-id="Analysis-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Analysis</a></span></li></ul></li></ul></li><li><span><a href="#Reference" data-toc-modified-id="Reference-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Reference</a></span></li></ul></div>

# Group information

| ID       | Name        | 
|:--------:|:------------|
| 20127323 | Võ Nhật Tân |
| 20127447 | Ngô Đức Bảo |


# Dataset information

## Subject of the data
- Heart failure is a common event caused by cardiovascular diseases (CVDs), which are the main reason leading to death globally. As a result, people with cardiovascular disease or who are at high cardiovascular risk (due to the presence of one or more risk factors such as hypertension, diabetes, hyperlipidaemia or already established disease) need early detection and management.

## Source of the data and how this dataset collected?
- This dataset is collected by FEDESORIANO in the Kaggle. It can be found in that link: [Dataset Link](https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction)
- This dataset was created by combining different datasets already available independently but not combined before. In this dataset, 5 heart datasets are combined over 11 common features which makes it the largest heart disease dataset available so far for research purposes. The five datasets used for its curation are:

    * ***Cleveland***: 303 observations
    * ***Hungarian***: 294 observations
    * ***Switzerland***: 123 observations
    * ***Long Beach VA***: 200 observations
    * ***Stalog (Heart) Data Set***: 270 observations
    
    
        ==> Total: 1190 observations
        Duplicated: 272 observations
        Final dataset: 918 observations

- Every dataset used can be found under the Index of heart disease datasets from UCI Machine Learning Repository on the following link: https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/

## Copyright
- The data is free for sharing, creating and adapting. This license can be seen in that [License Link](https://opendatacommons.org/licenses/odbl/1-0/).
- The creators of this dataset:
    1. Hungarian Institute of Cardiology. Budapest: Andras Janosi, M.D.
    2. University Hospital, Zurich, Switzerland: William Steinbrunn, M.D.
    3. University Hospital, Basel, Switzerland: Matthias Pfisterer, M.D.
    4. V.A. Medical Center, Long Beach and Cleveland Clinic Foundation: Robert Detrano, M.D., Ph.D.
- Donor:
    * David W. Aha (aha '@' ics.uci.edu) (714) 856-8779


# Data exploration

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Load the dataset

In [2]:
heart_df = pd.read_csv('heart.csv')
heart_df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

## Get size of the dataset

In [3]:
row_size, col_size = heart_df.shape
print(f'A number of column: {col_size}')
print(f'A number of row: {row_size}')

A number of column: 12
A number of row: 918


## Exploring the rows

### Meaning of each rows

- Each row represents the case which is recorded

### Check the duplicated row

In [4]:
count_duplicated_row = heart_df.duplicated().sum()
print(f"Count the duplicated rows: {count_duplicated_row}")

Count the duplicated rows: 0


* Dataset don't have the duplicated rows.

## Exploring the columns

### Meaning of the columns

|  | Attributre     | Description                                                                                          |
|:------:|:---------------|:-----------------------------------------------------------------------------------------------------|
| 1      | Age            | age of the patient [years]                                                                           |
| 2      | Sex            | sex of the patient [M: Male, F: Female]                                                              |
| 3      | ChestPainType  | chest pain type [TA: Typical Angina, ATA: Atypical Angina, NAP: Non-Anginal Pain, ASY: Asymptomatic] |
| 4      | RestingBP      | resting blood pressure [mm Hg]                                                                       |
| 5      | Cholesterol    | serum cholesterol [mm/dl]                                                                            |
| 6      | FastingBS      | fasting blood sugar [1: if FastingBS > 120 mg/dl, 0: otherwise]                                      |
| 7      | RestingECG     | resting electrocardiogram results [Normal, ST, LVH]                                                  |
| 8      | MaxHR          | maximum heart rate achieved [Numeric value between 60 and 202]                                       |
| 9      | ExerciseAngina | exercise-induced angina [Y: Yes, N: No]                                                              |
| 10     | Oldpeak        | oldpeak = ST [Numeric value measured in depression]                                                  |
| 11     | ST_Slope       | the slope of the peak exercise ST segment [Up: upsloping, Flat: flat, Down: downsloping]             |
| 12     | HeartDisease   | output class [1: heart disease, 0: Normal]                                                           |

### Datatype of each columns

In [5]:
heart_df.dtypes

Age                 int64
Sex                object
ChestPainType      object
RestingBP           int64
Cholesterol         int64
FastingBS           int64
RestingECG         object
MaxHR               int64
ExerciseAngina     object
Oldpeak           float64
ST_Slope           object
HeartDisease        int64
dtype: object

With the binary columns, such as `FastingBS`; `ExerciseAngina`; `HeartDisease`, we convert these attributes to boolean datatype.

In [6]:
#Checking the values of these columns
binary_cols = ['FastingBS', 'ExerciseAngina', 'HeartDisease']

for col in binary_cols:
    print(f"Value of {col} column: ", list(heart_df[col].unique()))

Value of FastingBS column:  [0, 1]
Value of ExerciseAngina column:  ['N', 'Y']
Value of HeartDisease column:  [0, 1]


In [7]:
#Convert it into boolean datatype
heart_df["FastingBS"] = heart_df["FastingBS"].astype(bool)
heart_df["ExerciseAngina"] = heart_df["ExerciseAngina"] == 'Y'
heart_df["HeartDisease"] = heart_df["HeartDisease"].astype(bool)

In [8]:
heart_df.dtypes

Age                 int64
Sex                object
ChestPainType      object
RestingBP           int64
Cholesterol         int64
FastingBS            bool
RestingECG         object
MaxHR               int64
ExerciseAngina       bool
Oldpeak           float64
ST_Slope           object
HeartDisease         bool
dtype: object

### Explore numeric column(s)

In [9]:
heart_num_df = heart_df.select_dtypes(include=[np.int64, np.float64])
heart_num_df.head()

Age  RestingBP  Cholesterol  MaxHR  Oldpeak
0   40        140          289    172      0.0
1   49        160          180    156      1.0
2   37        130          283     98      0.0
3   48        138          214    108      1.5
4   54        150          195    122      0.0

In [10]:
heart_num_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 918 entries, 0 to 917
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Age          918 non-null    int64  
 1   RestingBP    918 non-null    int64  
 2   Cholesterol  918 non-null    int64  
 3   MaxHR        918 non-null    int64  
 4   Oldpeak      918 non-null    float64
dtypes: float64(1), int64(4)
memory usage: 36.0 KB


There are 918 entries as same as the number of non-null values. So those numeric columns have no missing value!

In [11]:
heart_num_df.describe()

Age   RestingBP  Cholesterol       MaxHR     Oldpeak
count  918.000000  918.000000   918.000000  918.000000  918.000000
mean    53.510893  132.396514   198.799564  136.809368    0.887364
std      9.432617   18.514154   109.384145   25.460334    1.066570
min     28.000000    0.000000     0.000000   60.000000   -2.600000
25%     47.000000  120.000000   173.250000  120.000000    0.000000
50%     54.000000  130.000000   223.000000  138.000000    0.600000
75%     60.000000  140.000000   267.000000  156.000000    1.500000
max     77.000000  200.000000   603.000000  202.000000    6.200000

In [12]:
#Split into 10 bins to detect the abnormal values
for col_name in heart_num_df.columns:
    print(f'Column {col_name} - Value counts in each bins: ')
    print(heart_num_df[col_name].value_counts(bins=10, sort=False))
    print('-'*50)

Column Age - Value counts in each bins: 
(27.95, 32.9]     12
(32.9, 37.8]      37
(37.8, 42.7]      86
(42.7, 47.6]     104
(47.6, 52.5]     148
(52.5, 57.4]     201
(57.4, 62.3]     175
(62.3, 67.2]     101
(67.2, 72.1]      39
(72.1, 77.0]      15
Name: Age, dtype: int64
--------------------------------------------------
Column RestingBP - Value counts in each bins: 
(-0.201, 20.0]      1
(20.0, 40.0]        0
(40.0, 60.0]        0
(60.0, 80.0]        1
(80.0, 100.0]      26
(100.0, 120.0]    265
(120.0, 140.0]    405
(140.0, 160.0]    177
(160.0, 180.0]     35
(180.0, 200.0]      8
Name: RestingBP, dtype: int64
--------------------------------------------------
Column Cholesterol - Value counts in each bins: 
(-0.604, 60.3]    172
(60.3, 120.6]       6
(120.6, 180.9]     70
(180.9, 241.2]    319
(241.2, 301.5]    248
(301.5, 361.8]     81
(361.8, 422.1]     14
(422.1, 482.4]      3
(482.4, 542.7]      3
(542.7, 603.0]      2
Name: Cholesterol, dtype: int64
-------------------------

As we can see, `Age` and `MaxHR` columns have suitable values. The remaining columns (`RestingBP`, `Cholesterol`, `Oldpeak`) have some abnormal values: the value should to be the positive values but it contains zero number or the negative values.

In [13]:
#RestingBP
values_not_suit = heart_num_df[heart_num_df["RestingBP"] <= 20]["RestingBP"].to_list()
print(f'Values are not suitable in RestingBP: {set(values_not_suit)}')
print(f'Count records which contain values are not suitable in RestingBP: {len(values_not_suit)}')

Values are not suitable in RestingBP: {0}
Count records which contain values are not suitable in RestingBP: 1


* In the `RestingBP` columns, there is a record which has invalid value. Therefore, the `RestingBP` value of this record is seen as null.

In [14]:
#Cholesterol
values_not_suit = heart_num_df[heart_num_df["Cholesterol"] <= 60.3]["Cholesterol"].to_list()
print(f'Values are not suitable in Cholesterol: {set(values_not_suit)}')
print(f'Count records which contain values are not suitable in Cholesterol: {len(values_not_suit)}')

Values are not suitable in Cholesterol: {0}
Count records which contain values are not suitable in Cholesterol: 172


* As same as `RestingBP` columns, in the `Cholesterol` columns, there are 172 records which has invalid value, but this values are equal by 0. Therefore, the `Cholesterol` value of this record is seen as null.

In [15]:
#Oldpeak
values_not_suit = heart_num_df[heart_num_df["Oldpeak"] <= 0.04]["Oldpeak"].to_list()
print(f'Values are not suitable in Oldpeak: {set(values_not_suit)}')
print(f'Count records which contain values are not suitable in Oldpeak: {len(values_not_suit)}')

Values are not suitable in Oldpeak: {0.0, -0.9, -0.1, -0.7, -0.8, -0.5, -1.0, -2.0, -1.1, -2.6, -1.5}
Count records which contain values are not suitable in Oldpeak: 381


* The percentage of records which have unsuitable values is more higher. Therefore just delete this columns.

In [16]:
heart_df = heart_df.drop(columns=["Oldpeak"])

In [ ]:
#Plot the distribution of each numeric columns
numeric_cols_df = heart_df.select_dtypes(exclude=[object, bool])

fig, axs = plt.subplots(nrows = 2,ncols = 2, figsize = (16,10))

for i in range(numeric_cols_df.shape[-1]):

    ax = axs[i//2, i%2]
    
    ax1 = numeric_cols_df.iloc[:, i].plot(kind='hist', ax = ax)
    ax2 = numeric_cols_df.iloc[:, i].plot(kind='kde', ax=ax, secondary_y=True, grid=True)
    
    ax1.set_ylabel('Frequency')
    ax2.set_ylabel('Density')
    ax2.set_ylim(0)

    title = 'Distribution : ' + numeric_cols_df.columns[i]
    plt.title(title)
    
fig.tight_layout()

With the numberic columns, it is witnessed that the distribution is nearly symmetric.
Cholestrol has a bidmodal data distribution.

### Explore categorical column(s)

In [ ]:
heart_cat_df = heart_df.select_dtypes(include=[object, bool])
heart_cat_df.head()

In [ ]:
heart_cat_df.info()

There are 918 entries as same as the categories of non-null values. So those numeric columns have no missing value!

In [ ]:
heart_cat_df.describe().T

In [ ]:
for column in heart_cat_df.columns:
    print(heart_cat_df[column].value_counts())
    print('-----------------------------------')

#### Explore the binary attributes

In [ ]:
binary_cols_df = heart_cat_df.select_dtypes(include='bool')
binary_cols_df.head()

In [ ]:
fig, axs = plt.subplots(nrows = 2,ncols = 2,figsize = (16,10))
for i in range(len(binary_cols_df.columns)):
    plt.subplot(2,2,i+1)
    ax = sns.countplot(x=binary_cols_df.iloc[:, i], order=[True, False])
    ax.set(xlabel=None, ylabel=None)
    title = f'Value counts in {binary_cols_df.columns[i]} columns ' 
    plt.title(title)

axs[-1, -1].remove()

A distribution of values `FastingBS` is not asymmetric in comparation with the other columns. Specifically, the amount of fasting blood sugar, which is more than 120md/dl, less than that in the remaining group.

#### Explore object attribute

In [ ]:
obj_cols_df = heart_cat_df.select_dtypes(include='object')
obj_cols_df.head()

In [ ]:
fig, axs = plt.subplots(nrows = 2,ncols = 2,figsize = (16,10))
for i in range(len(obj_cols_df.columns)):
    plt.subplot(2,2,i+1)
    ax = sns.countplot(x=obj_cols_df.iloc[:, i])
    ax.set(xlabel=None, ylabel=None)
    title = f'Value counts in {obj_cols_df.columns[i]} columns ' 
    plt.title(title)

As we can see here:
- The number of records in each categorie group don't distribute equally.
    * `Sex`: the number of male is higher than the number of female.
    * `ChestPainType`: many people have Asymptomatic (ASY), rest of it are seen with clear symptoms (ATA, NAP, ASY, TA).
    * `RestingECG`: Normal group has more records than others group (ST and LVH).
    * `ST_Slope`: while Up and Flat group have the high number of records, Down group hit the low of that.

### Further explore

In numeric columns, are there any correlation?

In [ ]:
# First let just fill null (zero) value in Cholesterol and RestingBP with mean of its column
temp_col_df = numeric_cols_df.copy()

restingBP_not_null = temp_col_df[temp_col_df['RestingBP'] != 0]
restingBP_mean = restingBP_not_null['RestingBP'].mean()
cholesterol_not_null = temp_col_df[temp_col_df['Cholesterol'] != 0]
cholesterol_mean = cholesterol_not_null['Cholesterol'].mean()

temp_col_df['RestingBP'].replace(to_replace=0, value=restingBP_mean, inplace=True)
temp_col_df['Cholesterol'].replace(to_replace=0, value=cholesterol_mean, inplace=True)

temp_col_df = pd.concat([temp_col_df, binary_cols_df], axis=1)

sns.heatmap(temp_col_df.corr(), annot=True);

As we can see here:
* Mostly attributes has a positive correlation with other atributes, except `MaxHR` attribute.

In [ ]:
bins = heart_num_df['Age'].value_counts(bins=10, sort=False).index
temp_df = heart_df.copy()
temp_df['Age_binned'] = pd.cut(temp_df['Age'], bins)
ax = sns.countplot(data=temp_df, x='Age_binned', hue='Sex')
plt.xticks(rotation=45);

* Although distribution of value in `Male` columns is symmetric but the number of male is more than the number of female.

In [ ]:
consider_cols = ['RestingECG', 'ST_Slope', 'ChestPainType', 'FastingBS', 'ExerciseAngina']
fig, axs = plt.subplots(nrows = 3,ncols = 2, figsize = (16,15))

for i in range(len(consider_cols)):
    plt.subplot(3, 2, i+1)
    sns.countplot(data=heart_df, x=consider_cols[i], hue='HeartDisease', hue_order=[True, False])

axs[-1, -1].remove()

For almost attributes except `RestingECG` and `FastingBS`, we can clearly see the differences between the `HeartDisease` outcomes in each categorical value. However, there is less constrast in the outcomes of `HeartDisease` in the remaining attributes. Although the `FastingBS` is witnessed that trend, it doesn't have more information enough to fully conclude.

# Asking meaningful questions and answering them

In [ ]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
from sklearn.model_selection  import cross_val_score
from sklearn.metrics  import confusion_matrix
from sklearn.base import BaseEstimator, TransformerMixin
from mlxtend.frequent_patterns import apriori, association_rules
pd.set_option('max_colwidth', 1000)

## How to detect people who have heart disease by those symptoms?

### Purposes

* Detecting early heart failure is crucial for recognizing and curing it as fast as possible to reduce death rate. As the result, we need a model to detect it with the high accurancy and fast speed based on those symptoms 

- To answer this question, we need a machine learning model that learn our pattern then predict our future samples. We need to do this in several steps:
    * Step 1: Split the data to train set and test set.
    * Step 2: Pre-processing data, handle *null* values if needed.
    * Step 3: Evaluate several models on the train set using cross validation technique by taking accurancy as measure. 
    * Step 4: Evaluate the final best model on the test set.
    * Step 5: Retrain on full dataset to make the final model.

### Pre-processing

In [ ]:
heart_df.head()

Since numeric columns have *null* value, we will to fill those value with `SimpleImputer`. For all categorical columns, we will use `OneHotEncoder`

In [ ]:
X = heart_df.iloc[:, :-1]
y = heart_df.iloc[:, -1]

# Split the data: 80% train, 20% test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2022)

# Split num columns and category columns
num_cols = X.select_dtypes(exclude=[object, bool]).columns.tolist()
cate_cols = X.select_dtypes(include=[object, bool]).columns.tolist()

# Preprocess data
preprocessor = ColumnTransformer(transformers=[
    ('category', OneHotEncoder(handle_unknown='ignore'), cate_cols)
])

### Analysis

Now let's train and test the data on train set!

In [ ]:
full_model_logistic = Pipeline(steps=[
    ('preprocess', preprocessor),
    ('model', LogisticRegression(max_iter=1000, C=10, random_state=2022))
])

full_model_xgboost = Pipeline(steps=[
    ('preprocess', preprocessor),
    ('model', GradientBoostingClassifier(learning_rate=1.0, random_state=2022))
])

full_model_mlp = Pipeline(steps=[
    ('preprocess', preprocessor),
    ('model', MLPClassifier(hidden_layer_sizes=(100, ), max_iter=1000, random_state=2022))
])

# evaluate models on train set
print(f'Score on Logistic Regression: {cross_val_score(full_model_logistic, X_train, y_train, cv=5).mean()*100:.2f}%')
print(f'Score on Gradient Boosting: {cross_val_score(full_model_xgboost, X_train, y_train, cv=5).mean()*100:.2f}%')
print(f'Score on MLP Classifier: {cross_val_score(full_model_mlp, X_train, y_train, cv=5).mean()*100:.2f}%')

Look like Logistic Regression has highest score!  
We will perform our prediction on test set with Logistic Regression.

In [ ]:
# train on train set and predict on test set through best model
full_model_logistic.fit(X_train, y_train)
y_pred = full_model_logistic.predict(X_test)

# evaluate model prediction
tn, fp, fn, tp = confusion_matrix(y_pred=y_pred, y_true=y_test).ravel()

print(f'Accuracy: {((tp + tn) / (tn + fp + fn + tp) * 100):.2f}%')
print(f'Sensitivity (Recall): {(tp / (tp + fn) * 100):.2f}%')
print(f'Specificity: {(tn / (tn + fp) * 100):.2f}%')
print(f'Precision: {(tp / (tp + fp) * 100):.2f}%')

Not a bad result!

## Which factors leading to the heart disease?

### Purposes

* Exploring this question helps to early warn to the patient when the symtoms commence. Thanks to it, the patient can be make the better decisons. 

* To solve this question, we need to mine the items which are frequence in this database. To do that, we need to follow these steps:
    1. Convert all the numeric attributes to the categories.
        * Idea: for each numeric attributes, we divide into small range and then assign the labels to each range. With each element belongs to this range, we assign this range's label to this element.
        * Dividing the range and assigning it to the labels need to suitable, especially correction in the medicine study.  
        * To cover this, we categorize based on the `A Fuzzy Expert System for Heart Disease Diagnosis` study. It can be found in this [link](https://www.researchgate.net/publication/44260568_A_Fuzzy_Expert_System_for_Heart_Disease_Diagnosis).
    2. Encode these categorical features as a one-hot bits array.
    3. Mining the frequency items using Apriori algorithms.
    4. Creates the rules from the frequency items set.

### Pre-processing

* Pre-processing with 2 phases:
    1. Convert the numeric attributes to categorical attribute
    2. Convert it into the bit array.
    
    
* **Notice**: In step 1, we divide by below conditions (mentioned in the [study](https://www.researchgate.net/publication/44260568_A_Fuzzy_Expert_System_for_Heart_Disease_Diagnosis) above)
    + For the Age columns (Age):
        | RANGE   | FUZZY SETS |
        |:-------:|:----------:|
        | < 38    | Young      |
        | 33 - 45 | Mild       |
        | 40 - 58 | Old        |
        | > 52    | Very old   |
        
    + For the RestingBP columns (Blood Sugar (Diabetes)):
        | RANGE     | FUZZY SETS |
        |:---------:|:----------:|
        | > 120     | True       |
        | Otherwise | False      |
        
    + For the Cholesterol columns (Cholesterol):
        | RANGE     | FUZZY SETS |
        |:---------:|:----------:|
        | < 197     | Low        |
        | 188 - 250 | Medium     |
        | 217 - 307 | High       |
        | > 281     | Very High  |
        
    + For the MaxHR columns (Maximum Heart Rate):
        | RANGE     | FUZZY SETS |
        |:---------:|:----------:|
        | < 141     | Low        |
        | 111 - 194 | Medium     |
        | > 152     |  High      |
        
        

In [ ]:
def AgeMapper(age_series):
    age_col = age_series.copy()
    
    age_col[age_series < 38] = 'Young'
    age_col[age_series.between(33, 45)] = 'Mild'
    age_col[age_series.between(40, 58)] = 'Old'
    age_col[age_series > 52] = 'Very_Old'
    return age_col
    
def RestingBPMapper(resting_series):
    resting_col = resting_series.copy()
    resting_col = resting_col > 120
    return resting_col
    
def CholesterolMapper(cholesterol_series):
    cholesterol_col = cholesterol_series.copy()
    cholesterol_col[cholesterol_series < 197] = 'Low'
    cholesterol_col[cholesterol_series.between(188, 250)] = 'Medium'
    cholesterol_col[cholesterol_series.between(217, 307)] = 'High'
    cholesterol_col[cholesterol_series > 281] = 'Very_High'
    return cholesterol_col
    
def MaxHRMapper(maxHR_series):
    maxHR_col = maxHR_series.copy()
    maxHR_col[maxHR_series < 141] = 'Low'
    maxHR_col[maxHR_series.between(111, 194)] = 'Medium'
    maxHR_col[maxHR_series > 152] = 'High'
    return maxHR_col

In [ ]:
#Copy new dataframe.
heart_cate = heart_df.copy()

heart_cate['Age'] = AgeMapper(heart_cate['Age'])
heart_cate['RestingBP'] = RestingBPMapper(heart_cate['RestingBP'])
heart_cate['Cholesterol'] = CholesterolMapper(heart_cate['Cholesterol'])
heart_cate['MaxHR'] = MaxHRMapper(heart_cate['MaxHR'])
heart_cate[heart_cate.select_dtypes(bool).columns] = heart_cate[heart_cate.select_dtypes(bool).columns].astype(str)

In [ ]:
heart_cate.dtypes

In [ ]:
heart_cate.head()

### Analysis

In [ ]:
enc = OneHotEncoder(handle_unknown='ignore')
arr = enc.fit_transform(heart_cate).toarray().astype(bool)

heart_cate_one_hot_df =\
        pd.DataFrame(arr, columns=enc.get_feature_names(heart_cate.columns))

heart_cate_one_hot_df

In [ ]:
frequent_itemsets = apriori(heart_cate_one_hot_df, min_support = 0.2, use_colnames = True)
rules_df = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.9)
mask_result = rules_df["consequents"].apply(lambda x: 'HeartDisease_True' in x)

rules_df = rules_df[mask_result].sort_values(by=['confidence', 'support'], ascending = False)
rules_df = rules_df.reset_index(drop=True)
rules_df.loc[:5, ['antecedents', 'consequents', 'confidence', 'support' ]] 

* **Note:** The table above provides the information about the factors leading to Heart Disease with min_support = 0.2 and confidence is over 90%. The factor is an array of item which formated with `'symtom'_'value'`:
    + `systom` is the name of the columns
    + `value` is the unique value in this columns.

# Reference

1. Adeli, A., & Neshat, M. (2010, March). A Fuzzy Expert System for Heart Disease Diagnosis. Proceedings of the International MultiConference of Engineers and Computer Scientists 2010, I. https://www.researchgate.net/publication/44260568_A_Fuzzy_Expert_System_for_Heart_Disease_Diagnosis

2. (n.d.). Heart Failure Prediction Dataset. Kaggle. <br>
https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction